src/PHPCR/NodeInterface.php

```diff
@@ -15,13 +15,15 @@
 /**
  * The Node interface represents a node in a workspace.
  *
- * The \Traversable interface enables the implementation to be addressed with
+ * The Traversable interface enables the implementation to be addressed with
  * <b>foreach</b>. Nodes have to implement either \IteratorAggregate or
  * \Iterator.
  * The iterator is equivalent to <b>getNodes()</b> with no filter, returning
- * a list of all child nodes. Keys are the node names, values the node
+ * a list of all child nodes. Keys are the child node names, values the node
  * instances.
  *
+ * @extends Traversable<string, NodeInterface>
```

Comment:
Will it not always return the same class then `Traversable<string, static>` would be a better fit here.

Reply:
Okay not sure there can be different Nodes and child nodes if I understand correctly, because static shows this now:

<img width="1095" alt="Bildschirmfoto 2023-09-02 um 14 32 48" src="https://github.com/phpcr/phpcr/assets/1698337/82b2373c-7603-4e4b-8094-c92976a0b389">

Then `self` or `NodeInterface` is fine here if the childs can be different types.

Reply:
i guess that is up to the implementation. the spec only says that a NodeInterface resp PropertyInterface must be returned.

i think doing static would be wrong. maybe we could make it a template but not sure if a template can have a default value if none is further specified.

ideally, consumers would not need to know what concrete class is returned. because of that, i tend to say if it matters, implementations should overwrite the `@return` declaration as needed (if we would have actual return type declarations, narrowing them would also be allowed). or better, in the hack places where the actual type matters, do instanceof checks for what they received.

Reply:
As I know PHPStan does not yet support default types for Generics So fine currently keep it this way and the implementation can still override it.